PowerEditor/src/WinControls/ProjectPanel/Directory.cpp

```diff
@@ -0,0 +1,454 @@
+
+	// first, read directories, never filtered.
+	append(_path, TEXT("*.*"), true);
+
+	// then read files, by each filter
+	if (_filters.empty())
+	{
+		append(_path, TEXT("*.*"), false);
+	}
+	else
+	{
+		for (size_t i=0; i<_filters.size(); ++i)
+			append(_path, _filters[i], false);
+	}
+
+
+	readLastWriteTime(_lastWriteTime);
+
+	_wasRead = true;
+
+	return _exists;
+
+}
+
+bool Directory::readIfChanged(bool respectEmptyDirs)
+{
+	if (!_wasRead)
+	{
+		read();
+		return true;
+	}
+
+	if (respectEmptyDirs && _hideEmptyDirs)
+	{
+		if (!writeTimeHasChanged() && !containsDataChanged())
+			return false;
+	}
+	else
+	{
+		if (!writeTimeHasChanged())
+			return false;
+	}
+
+	const Directory cmp(_path,_filters);
+	if (*this != cmp)
+	{
+		*this = cmp;
+		return true;
+	}
+	return false;
+}
+
+void Directory::append(const generic_string& path, const generic_string& filter, bool readDirs)
+{
+	const bool readFiles = !readDirs;
+
+	generic_string searchPath = _path + TEXT("\\") + filter;
+
+	WIN32_FIND_DATA fd;
+
+	HANDLE hFind = FindFirstFileEx(searchPath.c_str(), FindExInfoBasic, &fd, FindExSearchNameMatch, NULL, 0);
+
+	if (hFind == INVALID_HANDLE_VALUE)
+		return;
+
+	_exists = true;
+
+	while (hFind != INVALID_HANDLE_VALUE)
+	{
+		const generic_string file(fd.cFileName);
+
+		if ((fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) != 0)
+		{
+			if (file == TEXT(".") || file == TEXT(".."))
+			{
+				if (!FindNextFile(hFind, &fd))
+					break;
+				continue;
+			}
+			if (readDirs)
+			{
+				if (_hideEmptyDirs)
+				{
+					generic_string childPath = path + TEXT("\\") + file;
+					if (!containsData(childPath))
+					{
+						_invisibleDirs.push_back(file);
+						if (!FindNextFile(hFind, &fd))
+							break;
+						continue;
+					}
+				}
+				_dirs.insert(file);
+			}
+		}
+		else
+		{
+			if (readFiles)
+				_files.insert(file);
+		}
+
+		if (!FindNextFile(hFind, &fd))
+			break;
+	}
+
+	if (hFind != INVALID_HANDLE_VALUE)
+		FindClose(hFind);
+
+
+}
+
+
+bool Directory::containsDataChanged() const
+{
+	for (auto it=_dirs.begin(); it != _dirs.end(); ++it)
+	{
+		generic_string childPath = _path + TEXT("\\") + *it;
+		if (!containsData(childPath))
+			return true;
+	}
+	for (auto it=_invisibleDirs.begin(); it != _invisibleDirs.end(); ++it)
+	{
+		generic_string childPath = _path + TEXT("\\") + *it;
+		if (containsData(childPath))
+			return true;
+	}
+	return false;
+}
+
+// containsData() checks, if a directory or its subdirs contain any data which match the filters.
+bool Directory::containsData(const generic_string& path) const
+{
+	if (_filters.empty())
+		return containsData(path, TEXT("*.*"));
+
+	for (size_t i=0; i<_filters.size(); ++i)
+	{
+		if (containsData(path, _filters[i]))
+			return true;
+	}
+	return false;
+}
+
+// containsData() recursive call
+bool Directory::containsData(const generic_string& path, const generic_string& filter) const
+{
+	generic_string searchPath = path + TEXT("\\") + filter;
+	generic_string childPath;
```

Comment:
This isn't `C89`, there's no need to declare variables at the beginning of scope! */grumpy*

Declare it on line 270, where it's initialized. I don't expect that to be controversial, as it makes the code easier to read.